chore: Add documentation for Syntax and coding example. (#9)
ZEISS-PiWeb/PiWeb-Calculated-Characteristics: README.md

|-|:-|

# Overview

- [Introduction](#introduction)
- [Installation](#installation)
- [Formula Syntax](#formulasyntax)
- [How To](#howto)

<a id="markdown-introduction" name="introduction"></a>
# Introduction

> Calculated characteristics in PiWeb are characteristics without dedicated measured values. Furthermore their values are calculated by formulas. These formulas allow referencing values of other characteristics which can be regulary measured or even calculated too. Using calculated characteristics allows to provide precalculated values for reporting without the need to write system expressions over and over again.

Reviewer: Ich würde es so schreiben,

```suggestion
Calculated characteristics in PiWeb are characteristics without their own measured values. Instead, their values are calculated by formulas. These formulas allow referencing values of other characteristics that can be measured regularly or calculated as well. Using calculated characteristics makes it possible to provide pre-calculated values for reporting without having to write system expressions over and over again.
```

# Installation

The **PiWeb Calculated Characteristics library** is available via NuGet.

Get it at [NuGet.org](https://www.nuget.org/packages/Zeiss.PiWeb.CalculatedCharacteristics/).

```cmd
PM> Install-Package Zeiss.PiWeb.CalculatedCharacteristics
```

Or compile the library by yourself. Requirements:

* Microsoft Visual Studio 2019
* Microsoft .NET Standard 2.1 or .NET5

<a id="markdown-formulasyntax" name="formulasyntax"></a>
# Formula Syntax

## Grammar

| node |   |
| --- | --- |
| expression | *expression* '+' *term* \| *expression* '-' *term* \| *term* |
| term | *term* '\*' *factor* \| *term* '/' *factor* \| *factor* |
| factor | [+-] ( *NUMBER* \| *function* \| '"' *IDENT* '"' \| '(' *expression* ')' \| '{' *path* '}' ) |
| function | *IDENT* '(' [*argumentlist*] ')' |
| argumentlist | *expression* \| *expression* '[,;]' *argumentlist* |
| path | *pathsegment* \| *pathsegment* '/' *path* \| *pathsegment* '(' *KEY* ')' |
| pathsegment | *IDENT* \| '"' *IDENT* '"' |
| NUMBER | any floating point value (with "." as decimal separator |
| IDENT | any sequence of characters |
| KEY | any integer in range [0,65535] |

## Characteristic Paths

> Characteristics are always placed in curly bracets. Per default the path is provided relative to the formula owners parent.

Reviewer: ```suggestion
Characteristics are always enclosed in curly brackets. By default, the path is specified relative to the parent origin of the formula.
```

### Escaping

> Part and characteristic names with characters `( ) { } " \ /` are supported using escaping. Escaping is possible by

Reviewer: ```suggestion
Part and characteristic names that contain following characters `( ) { } " \ /` can be escaped. Escaping is possible by
```

> 1. leading the character to escape by the escape character `\`

Reviewer: ```suggestion
1. Set the escape character `\` in front of the character to escape
```

> 2. encapsule the name by `"`. Escaping `"` and `\` within the escaped name can be done with the escape character `\`.

Reviewer: ```suggestion
2. Enclose the name by `"`. Escaping `"` and `\` within the escaped name can be done with the escape character `\`.
```

The following list shows some examples defining characteristic paths:

- {abc} -> References to characteristic ***abc*** of the same parent part or characteristic

> - {abc(20)} -> References to characteristic attribute with key ***20*** of charatceristic ***abc***

Reviewer: ```suggestion
- {abc(20)} -> References to characteristic attribute with key ***20*** of characteristic ***abc***
```

- {abc\(20\)} -> References to characteristic ***abc(20)***
- {"abc(20)"} -> References to characteristic ***abc(20)***
- {abc \"20\"} -> References to characteristic ***abc"20"***
- {"abc \"20\""} -> References to characteristic ***abc"20"***

> - {"abc \"20\""(20)} -> References to characteristic attribute with key ***20*** of charatceristic ***abc"20"***

Reviewer: ```suggestion
- {"abc \"20\""(20)} -> References to characteristic attribute with key ***20*** of characteristic ***abc"20"***
```

> - {../xyz/abc} -> References to characteristic ***abc*** of part or characrteristic ***xyz***

Reviewer: ```suggestion
- {../xyz/abc} -> References to characteristic ***abc*** of part or characteristic ***xyz***
```

            object GetValueFromAttribute(PathInformationDto path, ushort key)
            {
                return structureMap.TryGetValue(path, out var item)
                    ? item.Entity.Attributes?.FirstOrDefault(a => a.Key == key)?.Value
                    : null;
            }

            // create the calculators (once created we can reuse it)
            var factory = new AttributeBasedMathInterpreterFactory(GetValueFromAttribute, ChildPathsHandler);
            var char1Calculator = factory.GetCharacteristicCalculator(char1.Path);
            var char2Calculator = factory.GetCharacteristicCalculator(char2.Path);

            // should output "Result of Char1: 3"
            Console.Write("Result of Char1: ");
            Console.WriteLine(char1Calculator.GetResult(GetMeasurementValueForPath, (path, key, _) => GetValueFromAttribute(path, key)));

            // should output "Result of Char2: 8"
            Console.Write("Result of Char2: ");
            Console.WriteLine(char2Calculator.GetResult(GetMeasurementValueForPath, (path, key, _) => GetValueFromAttribute(path, key)));
        }
    }
}
````

> The type AttributeBasedMathInterpreterFactory reads the formula from the predefined characterstic attribute. If another formula source required, inherit from SimpleMathInterpreterFactory and override method GetFormula.

Reviewer: ```suggestion
The type `AttributeBasedMathInterpreterFactory` reads the formula from the predefined characteristic attribute. If another formula source is required, inherit from `SimpleMathInterpreterFactory` and override method `GetFormula`.
```
